scapy/contrib/ife.py

```diff
@@ -0,0 +1,110 @@
+#! /usr/bin/env python
+#
+# scapy.contrib.description = ForCES Inter-FE LFB type (IFE)
+# scapy.contrib.status = loads
+
+"""
+    IFE - ForCES Inter-FE LFB type
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    :author:    Alexander Aring, aring@mojatatu.com
+    :license:   GPLv2
+
+        This module is free software; you can redistribute it and/or
+        modify it under the terms of the GNU General Public License
+        as published by the Free Software Foundation; either version 2
+        of the License, or (at your option) any later version.
+
+        This module is distributed in the hope that it will be useful,
+        but WITHOUT ANY WARRANTY; without even the implied warranty of
+        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+        GNU General Public License for more details.
+
+    :description:
+
+        This module provides Scapy layers for the IFE protocol.
+
+        normative references:
+            - RFC 8013
+              Forwarding and Control Element Separation (ForCES)
+              Inter-FE Logical Functional Block (LFB)
+              https://tools.ietf.org/html/rfc8013
+"""
+
+from scapy.data import ETHER_TYPES
+from scapy.packet import Packet, bind_layers
+from scapy.fields import FieldLenField, PacketListField, IntField, \
+    ShortField, PadField, StrNullField
+from scapy.layers.l2 import Ether
+
+ETH_P_IFE = 0xed3e
+ETHER_TYPES['IFE'] = ETH_P_IFE
+
+
+class IFETlv(Packet):
+    """
+    Parent Class interhit by all ForCES TLV strucutures
+    """
+    pass
+
+
+class IFE(Packet):
+    """
+    Main IFE Packet Class
+    """
+    name = "IFE"
+
+    META_SKBMARK = 1
+    META_HASHID = 2
+    META_PRIO = 3
+    META_QMAP = 4
+    META_TCINDEX = 5
+
+    fields_desc = [FieldLenField("mdlen", None, length_of="tlvs",
+                                 adjust=lambda pkt, x: x+2),
+                   PacketListField("tlvs", None, IFETlv),
+                  ]
+
+
+def gen_ife_tlv_fields(ife_type, value_cls, value):
+    """
+    Runtime generation of fields_desc for IFE TLV classes
+    """
+    return [ShortField("type", ife_type),
+            FieldLenField("length", None, length_of="value",
+                          adjust=lambda pkt, x: x+4),
+            PadField(value_cls("value", value), 4, padwith=b'\x00'),
```

Comment:
That could be replaced by a `MultipleTypeField`. (You can still use PadField around) Example of usage: https://github.com/secdev/scapy/blob/a58e1b90a704c394216a0b5a864a50931754bdf7/scapy/contrib/socks.py#L126-L136